File: raspberry-pi-agent/src/agent.ts
import axios from 'axios';
import fs from 'fs/promises';
import path from 'path';
import { fileURLToPath } from 'url';
import { GoogleGenAI } from "@google/genai";
import {
    DeviceConfig,
    SensorConfig,
    ISensorDriver,
    ReadingPayload,
    AgentCommand,
    AgentState,
} from './types.js';

// --- Path Configuration ---
const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);
const CONFIG_PATH = path.join(__dirname, '..', 'config.json');

// --- Timers ---
const CONFIG_POLL_INTERVAL = 60000; // 1 minute
const SENSOR_READ_INTERVAL = 1000;  // Check every second if a sensor needs to be read
const COMMAND_POLL_INTERVAL = 5000; // 5 seconds

// Local config file structure
interface LocalConfig {
    server: { base_url: string };
    device: { id: string; token: string };
}

class Agent {
    private state: AgentState = AgentState.INITIALIZING;
    private config: DeviceConfig | null = null;
    private driverInstances: Map<string, ISensorDriver> = new Map();
    private lastReadTimes: Map<string, number> = new Map();
    private globalReadFrequencySeconds?: number;

    // Properties from local config
    private apiBaseUrl: string = '';
    private deviceId: string = '';
    private authToken: string = '';

    constructor(localConfig: LocalConfig) {
        this.apiBaseUrl = `${localConfig.server.base_url}/api`;
        this.deviceId = localConfig.device.id;
        this.authToken = localConfig.device.token;

        console.log(`🚀 ORION Agent Başlatılıyor... Cihaz ID: ${this.deviceId}`);
        this.setState(AgentState.INITIALIZING);
    }

    private setState(newState: AgentState) {
        if (this.state !== newState) {
            this.state = newState;
            console.log(`Durum Değişikliği: ${this.state}`);
        }
    }

    public async start() {
        await this.fetchConfig();
        
        setInterval(() => this.fetchConfig(), CONFIG_POLL_INTERVAL);
        setInterval(() => this.processSensors(), SENSOR_READ_INTERVAL);
        setInterval(() => this.checkForCommands(), COMMAND_POLL_INTERVAL);
    }

    private async fetchConfig() {
        console.log("🔄 Yapılandırma sunucudan alınıyor...");
        this.setState(AgentState.CONFIGURING);
        try {
            const response = await axios.get<DeviceConfig>(`${this.apiBaseUrl}/config/${this.deviceId}`, {
                headers: { 'Authorization': `Token ${this.authToken}` },
            });
            this.config = response.data;
            this.globalReadFrequencySeconds = this.config.global_read_frequency_seconds;
            console.log(`✅ Yapılandırma alındı: ${this.config.sensors.length} sensör, ${this.config.cameras.length} kamera. Global frekans: ${this.globalReadFrequencySeconds || 'devre dışı'}`);
            this.setState(AgentState.ONLINE);
        } catch (error) {
            this.setState(AgentState.OFFLINE);
            if (axios.isAxiosError(error)) {
                console.error(`❌ Yapılandırma alınamadı: ${error.message} (Durum: ${error.response?.status})`);
            } else {
                console.error('❌ Yapılandırma alınırken beklenmedik bir hata oluştu:', error);
            }
        }
    }

    private async processSensors() {
        if (this.state !== AgentState.ONLINE || !this.config) {
            return;
        }

        const now = Date.now();
        const sensorsToRead: SensorConfig[] = [];

        for (const sensorConfig of this.config.sensors) {
            if (!sensorConfig.is_active || sensorConfig.interface === 'virtual' || sensorConfig.type === 'Kar Yüksekliği') {
                continue;
            }

            const lastRead = this.lastReadTimes.get(sensorConfig.id) || 0;
            const individualFrequencySeconds = sensorConfig.read_frequency || 300;
            
            const effectiveFrequencySeconds = (this.globalReadFrequencySeconds && this.globalReadFrequencySeconds > 0)
                ? this.globalReadFrequencySeconds
                : individualFrequencySeconds;
            
            if (now - lastRead >= effectiveFrequencySeconds * 1000) {
                sensorsToRead.push(sensorConfig);
                this.lastReadTimes.set(sensorConfig.id, now); // Mark as "reading" to prevent re-triggering
            }
        }
        
        if (sensorsToRead.length > 0) {
            console.log(`🌡️ Okunacak ${sensorsToRead.length} sensör bulundu...`);
            await Promise.all(sensorsToRead.map(async (sensorConfig) => {
                console.log(`   - Okunuyor: ${sensorConfig.name} (ID: ${sensorConfig.id})`);
                try {
                    const driver = await this.loadDriver(sensorConfig.parser_config.driver);
                    if (!driver) {
                        console.error(`     -> HATA: Sürücü bulunamadı: ${sensorConfig.parser_config.driver}`);
                        return;
                    }
                    const reading = await driver.read(sensorConfig.config);
                    if (reading) {
                        console.log(`     -> Okunan Değer:`, reading);
                        await this.sendReading({ sensor: sensorConfig.id, value: reading });
                    } else {
                        console.warn(`     -> UYARI: ${sensorConfig.name} sensöründen veri okunamadı (null döndü).`);
                    }
                } catch (error) {
                    console.error(`     -> HATA: ${sensorConfig.name} sensörü okunurken hata oluştu:`, error);
                }
            }));
        }
    }

    private async loadDriver(driverName: string): Promise<ISensorDriver | null> {
        if (this.driverInstances.has(driverName)) {
            return this.driverInstances.get(driverName)!;
        }
        try {
            // Drivers are compiled to .js files in the dist directory
            const driverPath = `./drivers/${driverName}.driver.js`;
            const driverModule = await import(driverPath);
            const driverInstance: ISensorDriver = new driverModule.default();
            this.driverInstances.set(driverName, driverInstance);
            return driverInstance;
        } catch (error) {
            console.error(`Sürücü yüklenemedi: ${driverName}`, error);
            return null;
        }
    }

    private async sendReading(payload: ReadingPayload) {
        try {
            await axios.post(`${this.apiBaseUrl}/submit-reading`, payload, {
                headers: { 'Authorization': `Token ${this.authToken}` },
            });
            console.log(`     -> ✅ Değer sunucuya gönderildi (Sensör ID: ${payload.sensor})`);
        } catch (error) {
            if (axios.isAxiosError(error)) {
                console.error(`     -> ❌ Değer gönderilemedi: ${error.message}`);
            } else {
                console.error('     -> ❌ Değer gönderilirken beklenmedik hata:', error);
            }
        }
    }

    private async checkForCommands() {
        if (this.state !== AgentState.ONLINE) return;
        
        try {
            const response = await axios.get<AgentCommand[]>(`${this.apiBaseUrl}/commands/${this.deviceId}`, {
                headers: { 'Authorization': `Token ${this.authToken}` },
            });
            const commands = response.data;
            if (commands.length > 0) {
                console.log(`📩 ${commands.length} yeni komut alındı.`);
                for (const command of commands) {
                    await this.executeCommand(command);
                }
            }
        } catch (error) {
            if (axios.isAxiosError(error) && error.response?.status !== 404) {
                 console.error(`Komutlar kontrol edilirken hata: ${error.message}`);
            }
        }
    }

    private async executeCommand(command: AgentCommand) {
        console.log(`⚡ Komut yürütülüyor: ${command.command_type} (ID: ${command.id})`);
        let success = false;
        try {
            switch (command.command_type) {
                case 'CAPTURE_IMAGE':
                    success = await this.captureImage(command);
                    break;
                case 'ANALYZE_SNOW_DEPTH':
                    success = await this.analyzeSnowDepth(command);
                    break;
                case 'FORCE_READ_SENSOR':
                    success = await this.forceReadSensor(command);
                    break;
                default:
                    console.warn(`Bilinmeyen komut tipi: ${command.command_type}`);
                    break;
            }
            await this.updateCommandStatus(command.id, success ? 'complete' : 'fail');
        } catch (error) {
            console.error(`Komut yürütülürken hata oluştu (ID: ${command.id}):`, error);
            await this.updateCommandStatus(command.id, 'fail');
        }
    }
    
    private async updateCommandStatus(commandId: number, status: 'complete' | 'fail') {
        try {
            await axios.post(`${this.apiBaseUrl}/commands/${commandId}/${status}`, {}, {
                 headers: { 'Authorization': `Token ${this.authToken}` },
            });
            console.log(`✅ Komut durumu güncellendi: ID ${commandId} -> ${status}`);
        } catch (error) {
             console.error(`❌ Komut durumu güncellenemedi (ID: ${commandId}):`, error);
        }
    }

    private async captureImage(command: AgentCommand): Promise<boolean> {
        const { camera_id } = command.payload;
        if (!camera_id) {
            console.error("Capture image command is missing 'camera_id' in payload.");
            return false;
        }

        const cameraConfig = this.config?.cameras.find(c => c.id === camera_id);

        if (!cameraConfig) {
            console.error(`Fotoğraf çekilemedi: Kamera ID'si ${camera_id} yapılandırmada bulunamadı.`);
            return false;
        }

        const timestamp = new Date().toISOString().replace(/[:.]/g, '-');
        const filename = `${timestamp}_${this.deviceId}_${camera_id}.jpg`;
        const UPLOADS_DIR = path.join(__dirname, '..', 'uploads');
        const filepath = path.join(UPLOADS_DIR, filename);

        console.log(`📸 Fotoğraf çekiliyor: ${cameraConfig.name}...`);

        try {
            await fs.mkdir(UPLOADS_DIR, { recursive: true });
            
            // This is a mock capture. In a real scenario, you'd use a library
            // like `node-webcam` or execute a command-line tool like `fswebcam` or `ffmpeg`.
            // For example: `ffmpeg -i ${cameraConfig.rtsp_url} -vframes 1 ${filepath}`
            
            // Mock implementation: create a dummy file.
            await fs.writeFile(filepath, "dummy image data");
            console.log(`🖼️  Mock görüntü kaydedildi: ${filepath}`);

            // Read the file and send it as base64
            const imageBuffer = await fs.readFile(filepath);
            const base64Image = imageBuffer.toString('base64');
            
            await axios.post(`${this.apiBaseUrl}/cameras/${camera_id}/upload-photo`, {
                image: base64Image,
                filename: filename
            }, {
                headers: { 'Authorization': `Token ${this.authToken}` },
            });

            console.log(`🚀 Fotoğraf sunucuya yüklendi: ${filename}`);
            
            // Clean up local file after upload
            await fs.unlink(filepath);
            
            return true;

        } catch (error) {
            console.error(`HATA: Fotoğraf çekme veya yükleme başarısız oldu (Kamera: ${camera_id})`, error);
            return false;
        }
    }

    private async analyzeSnowDepth(command: AgentCommand): Promise<boolean> {
        const { camera_id, virtual_sensor_id } = command.payload;
        if (!camera_id) {
            console.error('Kar derinliği analizi için camera_id gerekli.');
            return false;
        }
        if (!virtual_sensor_id) {
            console.error('Kar derinliği analizi için virtual_sensor_id gerekli.');
            return false;
        }
        const cameraConfig = this.config?.cameras.find(c => c.id === camera_id);
        if (!cameraConfig) {
            console.error(`Analiz için kamera bulunamadı: ${camera_id}`);
            return false;
        }
        
        if (!process.env.API_KEY) {
            console.error('HATA: Gemini API anahtarı (API_KEY) ortam değişkenlerinde ayarlanmamış. Analiz yapılamıyor.');
            return false;
        }

        console.log(`❄️  Kar derinliği analizi başlatılıyor... Kamera: ${camera_id}`);
        
        const timestamp = new Date().toISOString().replace(/[:.]/g, '-');
        const filename = `ANALYSIS_${timestamp}_${camera_id}.jpg`;
        const UPLOADS_DIR = path.join(__dirname, '..', 'uploads');
        const filepath = path.join(UPLOADS_DIR, filename);

        try {
            // 1. Capture image
            await fs.mkdir(UPLOADS_DIR, { recursive: true });
            // MOCK: In a real scenario, use ffmpeg or similar
            await fs.writeFile(filepath, "dummy image data for analysis");
            console.log(`   -> Analiz için mock görüntü kaydedildi: ${filepath}`);

            // 2. Read image to base64
            const imageBuffer = await fs.readFile(filepath);
            const base64Image = imageBuffer.toString('base64');
            
            // 3. Call Gemini API
            console.log('   -> Gemini API ile analiz ediliyor...');
            const ai = new GoogleGenAI({ apiKey: process.env.API_KEY });
            const imagePart = {
                inlineData: {
                    mimeType: 'image/jpeg',
                    data: base64Image,
                },
            };
            const textPart = {
                text: "Bu görüntüdeki kar ölçüm cetveline göre karla kaplı en yüksek sayısal değer nedir? Cevabını sadece `{\"snow_depth_cm\": SAYI}` formatında bir JSON olarak ver.",
            };

            const response = await ai.models.generateContent({
                model: 'gemini-2.5-flash-image',
                contents: { parts: [imagePart, textPart] },
            });
            
            if (!response.text) {
                console.error('   -> HATA: Gemini API boş yanıt döndü.');
                return false;
            }

            const resultText = response.text.trim();
            console.log(`   -> Gemini Yanıtı: ${resultText}`);

            // 4. Parse response and send reading
            const resultJson = JSON.parse(resultText.replace(/```json/g, '').replace(/```/g, ''));
            const snowDepth = resultJson.snow_depth_cm;

            if (typeof snowDepth !== 'number') {
                console.error('   -> HATA: Gemini yanıtından sayısal bir kar yüksekliği değeri alınamadı.');
                return false;
            }
            
            console.log(`   -> Analiz Sonucu: ${snowDepth} cm`);
            
            await this.sendReading({
                sensor: virtual_sensor_id,
                value: { snow_depth_cm: snowDepth }
            });

            // 5. Upload analyzed image to server for verification
            await axios.post(`${this.apiBaseUrl}/analysis/upload-photo`, {
                cameraId: camera_id,
                image: base64Image,
                filename: filename
            }, {
                headers: { 'Authorization': `Token ${this.authToken}` },
            });
            console.log(`   -> Analiz görüntüsü sunucuya yüklendi: ${filename}`);

            return true;

        } catch (error) {
            console.error(`HATA: Kar derinliği analizi başarısız oldu (Kamera: ${camera_id})`, error);
            return false;
        } finally {
            // Clean up temp file
            try {
                await fs.unlink(filepath);
            } catch (unlinkError: any) {
                // Ignore error if file doesn't exist, but log others
                if (unlinkError.code !== 'ENOENT') {
                    console.warn(`Geçici analiz dosyası silinemedi: ${filepath}`, unlinkError);
                }
            }
        }
    }

    private async forceReadSensor(command: AgentCommand): Promise<boolean> {
        const { sensor_id } = command.payload;
        if (!sensor_id) {
            console.error("Force read komutu payload içinde 'sensor_id' içermiyor.");
            return false;
        }
    
        const sensorConfig = this.config?.sensors.find(s => s.id === sensor_id);
        if (!sensorConfig) {
            console.error(`Manuel okuma başarısız: Sensör ID ${sensor_id} mevcut cihaz yapılandırmasında bulunamadı.`);
            return false;
        }

        if (sensorConfig.interface === 'virtual') {
            console.warn(`Manuel okuma atlandı: ${sensorConfig.name} bir sanal sensördür ve bu yöntemle tetiklenemez.`);
            return true; // Return true as it's not a failure, just not applicable.
        }
    
        console.log(`⚡ Manuel okuma tetiklendi: ${sensorConfig.name}`);
    
        try {
            const driver = await this.loadDriver(sensorConfig.parser_config.driver);
            if (!driver) {
                console.error(`     -> HATA: Sürücü bulunamadı: ${sensorConfig.parser_config.driver}`);
                return false;
            }
            const reading = await driver.read(sensorConfig.config);
            if (reading) {
                console.log(`     -> Okunan Değer:`, reading);
                await this.sendReading({ sensor: sensorConfig.id, value: reading });
                return true; // Success
            } else {
                console.warn(`     -> UYARI: ${sensorConfig.name} sensöründen manuel okuma ile veri alınamadı.`);
                return false;
            }
        } catch (error) {
            console.error(`     -> HATA: ${sensorConfig.name} sensörü manuel okunurken hata oluştu:`, error);
            return false;
        }
    }
}


// --- Agent Startup ---
async function main() {
    try {
        console.log(`Yerel yapılandırma okunuyor: ${CONFIG_PATH}`);
        const configFile = await fs.readFile(CONFIG_PATH, 'utf-8');
        const localConfig: LocalConfig = JSON.parse(configFile);

        if (!localConfig.server?.base_url || !localConfig.device?.id || !localConfig.device?.token) {
            throw new Error("config.json dosyasında 'server.base_url', 'device.id', ve 'device.token' alanları zorunludur.");
        }

        const agent = new Agent(localConfig);
        agent.start().catch(error => {
            console.error("Agent çalışırken kritik bir hata oluştu:", error);
            // Fix: Cast process to 'any' to bypass TypeScript type error for 'exit'.
            (process as any).exit(1);
        });

    } catch (error) {
        console.error("Agent başlatılamadı. config.json dosyası okunamadı veya geçersiz.", error);
        // Fix: Cast process to 'any' to bypass TypeScript type error for 'exit'.
        (process as any).exit(1);
    }
}

main();

File: raspiagent-py/agent.py
import os
import time
import json
import requests
import importlib.util
import base64
import subprocess
from pathlib import Path
from dotenv import load_dotenv

# .env dosyasını yükle (API anahtarı gibi hassas veriler için)
load_dotenv() 

# --- Renk Tanımları ---
class Color:
    HEADER = '\033[95m'
    OKBLUE = '\033[94m'
    OKCYAN = '\033[96m'
    OKGREEN = '\033[92m'
    WARNING = '\033[93m'
    FAIL = '\033[91m'
    ENDC = '\033[0m'
    BOLD = '\033[1m'

# --- Zamanlayıcılar ---
CONFIG_POLL_INTERVAL = 60  # 1 dakika
SENSOR_READ_INTERVAL = 1      # Her saniye sensörleri kontrol et
COMMAND_POLL_INTERVAL = 5     # 5 saniye

# --- Yapılandırma ---
CONFIG_PATH = Path(__file__).parent / 'config.json'

class Agent:
    def __init__(self, local_config):
        self.state = "BAŞLATILIYOR"
        self.config = None
        self.driver_instances = {}
        self.last_read_times = {}
        self.global_read_frequency_seconds = 0

        self.api_base_url = f"{local_config['server']['base_url']}/api"
        self.device_id = local_config['device']['id']
        self.auth_token = local_config['device']['token']
        self.auth_header = {'Authorization': f'Token {self.auth_token}'}

        print(f"{Color.HEADER}🚀 ORION Agent Başlatılıyor... Cihaz ID: {self.device_id}{Color.ENDC}")
        self._set_state("BAŞLATILIYOR")

    def _set_state(self, new_state):
        if self.state != new_state:
            self.state = new_state
            print(f"{Color.OKBLUE}Durum Değişikliği: {self.state}{Color.ENDC}")

    def start(self):
        try:
            self.fetch_config()
            
            last_config_fetch = time.time()
            last_command_check = time.time()

            while True:
                if time.time() - last_config_fetch >= CONFIG_POLL_INTERVAL:
                    self.fetch_config()
                    last_config_fetch = time.time()

                if time.time() - last_command_check >= COMMAND_POLL_INTERVAL:
                    self.check_for_commands()
                    last_command_check = time.time()

                self.process_sensors()
                time.sleep(SENSOR_READ_INTERVAL)

        except KeyboardInterrupt:
            self.shutdown()
        except Exception as e:
            print(f"{Color.FAIL}Ana döngüde kritik hata: {e}{Color.ENDC}")
            self.shutdown()

    def shutdown(self):
        print(f"\n{Color.WARNING}🚫 Agent durduruluyor... Kaynaklar temizleniyor.{Color.ENDC}")
        print(f"{Color.OKGREEN}✅ Güvenli çıkış tamamlandı.{Color.ENDC}")

    def fetch_config(self):
        print("🔄 Yapılandırma sunucudan alınıyor...")
        self._set_state("YAPILANDIRILIYOR")
        try:
            response = requests.get(f"{self.api_base_url}/config/{self.device_id}", headers=self.auth_header, timeout=10)
            response.raise_for_status()
            self.config = response.json()
            self.global_read_frequency_seconds = self.config.get('global_read_frequency_seconds', 0)
            print(f"{Color.OKGREEN}✅ Yapılandırma alındı: {len(self.config['sensors'])} sensör, {len(self.config['cameras'])} kamera. Global frekans: {self.global_read_frequency_seconds}{Color.ENDC}")
            self._set_state("ÇEVRİMİÇİ")
        except requests.exceptions.RequestException as e:
            self._set_state("ÇEVRİMDIŞI")
            print(f"{Color.FAIL}❌ Yapılandırma alınamadı: {e}{Color.ENDC}")

    def get_effective_frequency(self, sensor_config):
        return self.global_read_frequency_seconds if self.global_read_frequency_seconds > 0 else sensor_config.get('read_frequency', 300)

    def process_sensors(self):
        if self.state != "ÇEVRİMİÇİ" or not self.config:
            return

        now = time.time()
        active_sensors = [s for s in self.config.get('sensors', []) if s.get('is_active') and s.get('interface') != 'virtual']
        
        for sensor_config in active_sensors:
            last_read = self.last_read_times.get(sensor_config['id'], 0)
            read_frequency = self.get_effective_frequency(sensor_config)

            if now - last_read >= read_frequency:
                print(f"[OKUMA BAŞLADI] {sensor_config['name']} (ID: {sensor_config['id']})")
                self.last_read_times[sensor_config['id']] = now

                try:
                    driver = self.load_driver(sensor_config['parser_config']['driver'])
                    if not driver:
                        print(f"  -> {Color.FAIL}HATA: Sürücü yüklenemedi: {sensor_config['parser_config']['driver']}{Color.ENDC}")
                        continue
                    
                    reading = driver.read(sensor_config['config'])
                    if reading is not None:
                        self.send_reading({'sensor': sensor_config['id'], 'value': reading})
                    else:
                        print(f"  -> {Color.WARNING}UYARI: {sensor_config['name']} sensöründen veri okunamadı.{Color.ENDC}")

                except Exception as e:
                    print(f"  -> {Color.FAIL}HATA ({sensor_config['parser_config']['driver']}): Sensör okunurken bir hata oluştu: {e}{Color.ENDC}")
                finally:
                    print(f"[OKUMA BİTTİ] {sensor_config['name']}")

    def load_driver(self, driver_name):
        if driver_name in self.driver_instances:
            return self.driver_instances[driver_name]
        try:
            driver_path = Path(__file__).parent / 'drivers' / f"{driver_name}_driver.py"
            spec = importlib.util.spec_from_file_location(driver_name, driver_path)
            module = importlib.util.module_from_spec(spec)
            spec.loader.exec_module(module)
            
            # Sürücü dosyasındaki sınıf adının CamelCase (örn: DfrobotUltDriver) olduğunu varsayıyoruz
            class_name = ''.join(word.capitalize() for word in driver_name.split('_')) + 'Driver'
            driver_class = getattr(module, class_name)
            
            instance = driver_class()
            self.driver_instances[driver_name] = instance
            return instance
        except Exception as e:
            print(f"{Color.FAIL}Sürücü {driver_name} yüklenemedi: {e}{Color.ENDC}")
            return None

    def send_reading(self, payload):
        try:
            response = requests.post(f"{self.api_base_url}/submit-reading", json=payload, headers=self.auth_header, timeout=10)
            response.raise_for_status()
            print(f"     -> ✅ Değer sunucuya gönderildi (Sensör ID: {payload['sensor']})")
        except requests.exceptions.RequestException as e:
            print(f"     -> {Color.FAIL}❌ Değer gönderilemedi: {e}{Color.ENDC}")
            self._set_state("ÇEVRİMDIŞI")

    def check_for_commands(self):
        if self.state != "ÇEVRİMİÇİ":
            return
        try:
            response = requests.get(f"{self.api_base_url}/commands/{self.device_id}", headers=self.auth_header, timeout=5)
            if response.status_code == 200:
                commands = response.json()
                if commands:
                    print(f"📩 {len(commands)} yeni komut alındı.")
                    for command in commands:
                        self.execute_command(command)
        except requests.exceptions.RequestException:
            # 404 veya timeout normal, hata basmaya gerek yok
            pass

    def execute_command(self, command):
        print(f"⚡ Komut yürütülüyor: {command['command_type']} (ID: {command['id']})")
        success = False
        try:
            if command['command_type'] == 'CAPTURE_IMAGE':
                success = self.capture_image(command['payload']['camera_id'])
            elif command['command_type'] == 'ANALYZE_SNOW_DEPTH':
                success = self.analyze_snow_depth(command['payload']['camera_id'], command['payload']['virtual_sensor_id'])
            # Diğer komutlar buraya eklenebilir
            self.update_command_status(command['id'], 'completed' if success else 'failed')
        except Exception as e:
            print(f"{Color.FAIL}Komut yürütülürken hata: {e}{Color.ENDC}")
            self.update_command_status(command['id'], 'failed')

    def update_command_status(self, command_id, status):
        try:
            requests.post(f"{self.api_base_url}/commands/{command_id}/{status}", headers=self.auth_header, timeout=5)
        except requests.exceptions.RequestException as e:
            print(f"{Color.FAIL}Komut durumu güncellenemedi: {e}{Color.ENDC}")

    def capture_image(self, camera_id):
        camera_config = next((c for c in self.config.get('cameras', []) if c['id'] == camera_id), None)
        if not camera_config or not camera_config.get('rtsp_url'):
            print(f"{Color.FAIL}Fotoğraf çekilemedi: Kamera (ID: {camera_id}) yapılandırması eksik.{Color.ENDC}")
            return False

        timestamp = time.strftime("%Y-%m-%dT%H-%M-%S", time.gmtime())
        filename = f"{timestamp}_{self.device_id}_{camera_id}.jpg"
        uploads_dir = Path(__file__).parent / 'uploads'
        uploads_dir.mkdir(exist_ok=True)
        filepath = uploads_dir / filename

        print(f"📸 Fotoğraf çekiliyor: {camera_config['name']}...")
        try:
            # Görüntüyü yakala, yeniden boyutlandır ve kalitesini ayarla
            subprocess.run(
                ['ffmpeg', '-i', camera_config['rtsp_url'], '-vframes', '1', '-vf', 'scale=1280:-1', '-q:v', '4', '-y', str(filepath)],
                check=True, capture_output=True, text=True
            )
            print(f"🖼️  Görüntü kaydedildi: {filepath}")

            with open(filepath, 'rb') as f:
                image_b64 = base64.b64encode(f.read()).decode('utf-8')
            
            payload = {'image': image_b64, 'filename': filename}
            requests.post(f"{self.api_base_url}/cameras/{camera_id}/upload-photo", json=payload, headers=self.auth_header)
            print(f"🚀 Fotoğraf sunucuya yüklendi: {filename}")
            
            os.remove(filepath)
            return True
        except (subprocess.CalledProcessError, requests.exceptions.RequestException) as e:
            print(f"{Color.FAIL}HATA: Fotoğraf çekme veya yükleme başarısız. {e}{Color.ENDC}")
            if isinstance(e, subprocess.CalledProcessError):
                print(f"FFMPEG Hata Çıktısı:\n{e.stderr}")
            return False

    def analyze_snow_depth(self, camera_id, virtual_sensor_id):
        # Bu fonksiyon, Gemini API kullanarak görüntü analizi yapar.
        # Python agent için benzer bir mantık uygulanabilir.
        # Şimdilik place holder olarak bırakıldı.
        print(f"{Color.WARNING}UYARI: 'analyze_snow_depth' fonksiyonu Python agent'ında henüz tam olarak implemente edilmedi.{Color.ENDC}")
        return True # Geçici olarak başarılı dönüyor


if __name__ == "__main__":
    try:
        with open(CONFIG_PATH, 'r') as f:
            local_config = json.load(f)
        agent = Agent(local_config)
        agent.start()
    except FileNotFoundError:
        print(f"{Color.FAIL}HATA: 'config.json' dosyası bulunamadı. Lütfen oluşturun.{Color.ENDC}")
    except json.JSONDecodeError:
        print(f"{Color.FAIL}HATA: 'config.json' dosyası geçersiz JSON formatında.{Color.ENDC}")
    except Exception as e:
        print(f"{Color.FAIL}Agent başlatılırken bir hata oluştu: {e}{Color.ENDC}")